### PR TITLE
Cache font vertices, GPU-side when possible

### PR DESCRIPTION
**Pull request by bavison**
This is quite a complex change, so I've formulated it as a series of dependent patches.

The aim is to reduce CPU/GPU bandwidth by keeping the vertices describing each string of font glyphs in a cache on the GPU side. Simple transformations (just translations in practice) are also performed on the GPU without needing to re-transfer the vertices.

At the moment, the main benefit is only available to OpenGL ES clients (such as the Raspberry Pi). However, there is a lesser benefit to be gained by the use of a CPU-side version of the cache, which is used in other cases, as well as in OpenGL ES cases where the selected transformation matrices rule out the use of hardware clipping.

I've benchmarked the typical rendering time of each frame on the main page of Confluence with RSS ticker and debug window enabled, on a non-overclocked Raspberry Pi:

```
      Before          After
      Mean   StdDev   Mean   StdDev  Confidence  Change
Time  20.5   3.8      18.2   3.8     100.0%      +12.5%
```

In FPS terms, 20.5 ms is 48.8 FPS, 18.2 ms is 54.9 FPS.

### CODE REVIEW

**Comment by bavison** [comment]
Here's a new rebase on master. I've also rolled in a fix for a heap corruption bug.

**Comment by fritsch** [comment]
Can you point out the heap corruption fix?

**Comment by bavison** [comment]
The bug was in CGUIFontTTFBase::DrawTextInternal(). It was hanging onto the reference returned by the appropriate instantiation of CGUIFontCache<Position, Value>::Lookup() across calls to CGUIFontTTFBase::GetCharacter(), which in exceptional circumstances (for example, I've only managed to replicate it when the video resolution was 1360x768) could cause the font cache to be flushed and thereby invalidate the reference. Now it does the lookup twice - once to see if the string is already in the cache, and a second time to find where to insert the vertex data after it has assembled the data it gleaned from GetCharacter().

The two instantiations of the CGUIFontCache template were added at different stages in the patch series, and the data types used also changed at various points, so it's not possible to identify one single point where the bugfix is applied.

**Comment by fritsch** [comment]
jenkins build this please

I am not sure whom to ask for review @popcornmix @topfs2 - if we want this code we should pull it early to v15 (now for example) :-)

This would reduce the distribution specific raspberry pi patch quite a lot.

**Comment by popcornmix** [comment]
This (minus the latest fix) has been in my test builds for almost a year.
It was in both Gotham and Helix builds for openelec/raspbmc.
It does improve performance on GLES platforms.

I think it should be pulled in now.

Perhaps @koying could test it on Android?

**Comment by koying** [comment]
Will do

**Comment by bavison** [comment]
If I'm reading that build failure right, the only problem is that the Windows build is failing to build a couple of add-ons because it can't find zlib.h. I'm scratching my head wondering how that has anything to do with this patch series - any ideas?

**Comment by fritsch** [comment]
It has nothing to do with it.

But will wait for confirmation from others.

2015-01-23 14:08 GMT+01:00 bavison <notifications@github.com>:

> If I'm reading that build failure right, the only problem is that the
> Windows build is failing to build a couple of add-ons because it can't find
> zlib.h. I'm scratching my head wondering how that has anything to do with
> this patch series - any ideas?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/4143#issuecomment-71190748>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by Memphiz** [comment]
its unrelated to this pr
@Jalle19 @Karlson2k do you have any idea how this problem could arise in ipvtsimple addon? It doesn't find zlib.h I verified (and also see it in the jenkins output) that the package was successfully downloaded and installed ... project/BuildDependencies/include/zlib.h exists in the workspace (verified it at this very moment - workspace will be ovewritten by next build though in case you wanna check it out yourself). Also it looks that the iptvsimple addon has "<AdditionalIncludeDirectories>..\..\..\..\project\BuildDependencies\include" set ... so whats going on here? (and why is it always zlib that makes problems *grr*)

Output of the build including error is here:

http://jenkins.kodi.tv/job/WIN-32/3290/consoleFull 

search for PVRIptvData.cpp

**Comment by Memphiz** [comment]
Ahhh just realised there is a seperate builddependencies scripts + packages in the xbmc-pvr-addons subdir here:

http://jenkins.kodi.tv/job/WIN-32/ws/project/BuildDependencies/tmp/xbmc-pvr-addons/project/BuildDependencies/include/

And indeed it is empty ... but there is no printout from this "sub" packaging system of pvr-addons ... most likely it was once more a network issue ...

Jep - just confirmed:

http://jenkins.kodi.tv/job/WIN-32/ws/project/BuildDependencies/tmp/xbmc-pvr-addons/project/BuildDependencies/downloads/zlib-vc100-1.2.5-lib.tar.bz2 is only 5kb in size.

@kib i have the feeling one of our mirrors has a broken zlib-vc100-1.2.5-lib.tar.bz2 should be in http://mirrors.kodi.tv/build-deps/win32/zlib-vc100-1.2.5-lib.tar.bz2 - can you somehow verify this?

**Comment by Jalle19** [comment]
@Memphiz I was just about to comment the same thing. The DownloadBuildDeps.bat for the addons is never run.

**Comment by Jalle19** [comment]
Actually it could be run, hard to say since the output is apparently silenced in buildpvraddons.bat

**Comment by Memphiz** [comment]
well it is for sure - other builds succeeded already ... if it was never run they would all have failed. I guess we just don't see the output of that ...

**Comment by Memphiz** [comment]
@Jalle19 just have a look in the workspace:

http://jenkins.kodi.tv/job/WIN-32/ws/project/BuildDependencies/tmp/xbmc-pvr-addons/project/BuildDependencies/downloads/

It was downloaded incomplete (only 5kb in size instead of 50kb) ... now every successive build will fail until i remove that file from the workspace so that it gets refetched...

**Comment by ace20022** [comment]
@Memphiz @Jalle19 in fact it is a website, you can click on the view button
Stating,e.g., "This content has been blocked because it does not comply with the acceptable usage policy."

**Comment by kib** [comment]
Another build failed because that same machine was not able to download something from github.
(See http://jenkins.kodi.tv/job/WIN-32/3283/console)

I am guessing this is network / firewall related on that machine. @zag2me could you take a look ?

**Comment by Memphiz** [comment]
yep "Unable to establish SSL connection." seen that too at some point i also informed zag who told me there was nothing changed. Its definitly dedicated to that slave and its network connection.

**Comment by Memphiz** [comment]
whiped out the workspace for now ... next build will refetch everything ...

**Comment by fritsch** [comment]
@Memphiz you want to trigger a manual build for windows?

**Comment by Memphiz** [comment]
why me? everybody can do so...

**Comment by fritsch** [comment]
I did not want to interrupt your tests, that's why I wanted to ask before doing another: jenkins build this please

**Comment by kib** [comment]
I did not remember how to do a windows build only in Jenkins...

**Comment by fritsch** [comment]
It's easy, just login to jenkins, click on the win32 build and start a new build with parameters on the left :-)

**Comment by kib** [comment]
@fritsch thx, I keep missing that option.

**Comment by fritsch** [comment]
Okay - button is green an "All is well". First come first serve -> next one that feels it can go in, should press it.

**Comment by popcornmix** [comment]
@koying any chance of testing this soon?

**Comment by bavison** [comment]
That should hopefully address jmarshallnz's minor comments.

**Comment by fritsch** [comment]
jenkins build this please

**Comment by koying** [comment]
Nothing to report, neither in menu or subs, which is expected I guess ;)
Are there specifics you want me test?

**Comment by popcornmix** [comment]
Scrolling text (e.g. RSS feed or "too long" filenames) will perhaps exercise more of the new code, but the expected behaviour is nothing to report.

**Comment by koying** [comment]
Scrolling text seems smoother indeed :)

**Comment by MartijnKaijser** [comment]
http://pastebin.com/NrZ30Xpu

**Comment by bavison** [comment]
CLangInfo::CRegion::CRegion - I don't think this one has anything to do with this PR.

CGUIFontCacheDynamicPosition - the only place where the default constructor is called is in CGUIFontTTFBase::DrawTextInternal. Throughout this function, there are conditional branches based on the hardwareClipping bool. When true, the object that was default-constructed is immediately initialised from another that uses a non-default constructor. When false, the object is not accessed, so the fact that it was uninitialised by the default constructor has no effect.

CScrollInfo::CScrollInfo unconditionally calls CScrollInfo::Reset, which sets m_widthValid=false, and m_widthValid guards all accesses to m_totalWidth, so initialising m_totalWidth would have no practical effect.

In summary, I don't think there are any bugs there. Of course, we could trivially remove the last two warnings at a slight overhead at runtime (and there would no longer be any point in having a default constructor for CGUIFontCacheDynamicPosition at all) - what are people's feelings on that?

**Comment by Memphiz** [comment]
This instantly crashes on ios:

http://pastebin.com/01NheJiL

"ERROR: 0:21: '' :  extension 'GL_OES_EGL_image_external' is not supported
ERROR: 0:24: 'm_samp0' : syntax error: syntax error" ... maybe this or something else is the issue?

**Comment by popcornmix** [comment]
@Memphiz sure it's this PR?
#6169 did edit m_samp0 in a shader and has:
#extension GL_OES_EGL_image_external : require

**Comment by koying** [comment]
Indeed, #6169 uses OES for android, and I doubt this one would use it.
But it is only used with the SM_TEXTURE_RGBA_OES (old) and SM_TEXTURE_RGBA_BOB_OES (new) guishaders, which I doubt ios would use.

I cannot check pastebin from work, unfortunately. Will do tonight.

**Comment by popcornmix** [comment]
I think the error in log is just trying to compile shader, not run it, and I think the bob shader is attempting to compile:
> 2015-01-28 01:13:19.304 Kodi[1487:841993] Debug Print: GUI Shader [guishader_frag_rgba_bob_oes.glsl] - Initialise failed

**Comment by koying** [comment]
And it crashes specifically on that? Strange...
"guishader_frag_rgba_oes.glsl" should fail also, and earlier.

**Comment by popcornmix** [comment]
Pretty sure #6169 causes the errors quoted by @Memphiz in log.
@Memphiz can you back put this PR and #6169 and see what the behaviour is?
The crash may be unrelated. Does @bavison has any thoughts.

**Comment by koying** [comment]
Could someone put the crashlog on something else than pastebin so that I can have a look, please ;)

**Comment by popcornmix** [comment]
http://xbmclogs.com/show.php?id=406178

**Comment by popcornmix** [comment]
http://xbmclogs.com/show.php?id=406178

**Comment by koying** [comment]
Crashes in CGUIFontTTFGL::LastEnd, which is definitely from this PR and unrelated to the shaders failed compilation.

**Comment by bavison** [comment]
The shader errors didn't ring any bells with me, but CGUIFontTTFGL::LastEnd -> glDrawElements -> memmove does look like it's connected to this PR. Any chance of someone who can test iOS doing a binary chop of the commits in this series to see where the crash was introduced? They should all be functional (though for a few in the middle, expect text clipping not to work).

**Comment by Memphiz** [comment]
binary chop? You mean git bisect?

**Comment by popcornmix** [comment]
git bisect will work, but as there are only 12 commits in this PR then manually chopping (git checkout <hash>) to find the first commit with the crash may be easier.

**Comment by Memphiz** [comment]
I will try to dedicate some time for this - maybe this evening.

**Comment by Memphiz** [comment]
typically and thanks to murphys law it was the last of the 12 commits:

https://github.com/xbmc/xbmc/commit/4237afaf7ed06301b3143b01a80092f1f501e526

**Comment by Memphiz** [comment]
it crashes on the very first call to https://github.com/xbmc/xbmc/commit/4237afaf7ed06301b3143b01a80092f1f501e526#diff-fb4b9c2fe7845ffd1274e5b5541a76d5R241 after startup ...

colLoc	GLint 2
character	size_t 0
posLoc	GLint 1
tex0Loc	GLint 0
count	size_t 8

**Comment by Memphiz** [comment]
@martijn - fyi - if we do alpha before this is solved - we might be prepared for extra support in the forum ;)

**Comment by bavison** [comment]
Doesn't look like I can add any more commits here now that it's been merged, soI've opened a fresh PR #6297 to address it. I've checked it on Linux and I'm told it now works on iOS as well.

**Comment by Montellese** [comment]
Running current master on win32 I came across some very odd font behaviour where it suddenly changes to a bunch of garbage, see e.g. https://dl.dropboxusercontent.com/content_link/bhnArS4L656QSjNxXs0IGiJt96DmmZ1FL1N0RiMVHKZRINMRKrtxsoRbZYW5LRhg?dl=1. Is it possible that this is the cause for that?

**Comment by popcornmix** [comment]
@Montellese your dropbox link doesn't work.
Any instructions on how to reproduce the error?

**Comment by Montellese** [comment]
@popcornmix: Sorry about that. Updated my last comment. Hope it works now.

TBH I don't know how to reproduce. I just saw it a few times now in totally different places while testing some stuff in Kodi.

**Comment by ace20022** [comment]
I can confirm the issue, but couldn't find a reliable way to reproduce.

**Comment by ace20022** [comment]
I now get it on every start, sample: https://www.dropbox.com/s/ed04pjnynccnwuy/test3_2.avi?dl=0
The sub menu fonts break.

**Comment by ace20022** [comment]
Since no one seems to care I vote for a revert.

**Comment by FernetMenta** [comment]
does a revert fix the issue on your system?

**Comment by ace20022** [comment]
> does a revert fix the issue on your system?

yes

**Comment by FernetMenta** [comment]
i agree, if not fixed very soon it should be reverted and fixed on a separate branch

**Comment by bavison** [comment]
It's a bit tricky when there isn't even a specific sequence of events that are known to trigger it. However, here are two possible approaches:

https://github.com/bavison/xbmc/tree/font_cache_directx_fix
- my current best guess of a fix

https://github.com/bavison/xbmc/tree/font_cache_directx_disable
- a fallback position, disables even the static font cache on Windows builds (the dynamic cache would needs a DirectX guru to write the equivalent of the OpenGL code, so is already disabled)

Can I have a volunteer to build and test the Windows versions of one or both of those?

**Comment by ace20022** [comment]
> It's a bit tricky when there isn't even a specific sequence of events that are known to trigger it. However, here are two possible approaches:
> 
> https://github.com/bavison/xbmc/tree/font_cache_directx_fix
> 
>     my current best guess of a fix

Thanks a lot! That seems to fix it. And now I can reproduce it, simply resizing the window (to trigger the realloc) ;)

**Comment by bavison** [comment]
That was fast! I'll raise a PR for that fix then, thanks

**Comment by FernetMenta** [comment]
@bavison thank you very much for your fast help

**Comment by Karlson2k** [comment]
@afedchin
As @bavison said https://github.com/xbmc/xbmc/pull/4143#issuecomment-72842077 win32 need DX equivalent of OpenGL code for dynamic font cache. Could you help us?

**Comment by afedchin** [comment]
@Karlson2k currently I'm working on DX11 version. I'll try to implement dx version of font cache with DX11 but I need a few explanations about of cache mechanics. I don't understand how it works exactly.

**Comment by afedchin** [comment]
@Karlson2k I've tried to implement DX equivalent. It randomly cause flicker label animation and sometimes cause disappearing label at the end of animation. It also tries to create an empty vertex buffers which not allowed in DirectX. If I disable the caching then issue disappears.

**Comment by afedchin** [comment]
@FernetMenta can you help to resolve an issue described below. I've debuged it and found what label which is disappeared have wrong translateY value in CTranslatedVertices struct. do you have any idea?

**Comment by FernetMenta** [comment]
@afedchin what steps are required to reproduce the issue? what branch should I use?

**Comment by afedchin** [comment]
@FernetMenta the issue appears in DX11 version Kodi where I've implemented support chaching font data at GPU. To reproduce the issue just move focus to different home menu items after some time sub-items will disappear. here is branch https://github.com/afedchin/xbmc/tree/dx11 here is test thread on the forum http://forum.kodi.tv/showthread.php?tid=218274 with latest version.

**Comment by FernetMenta** [comment]
@afedchin I compiled your branch and tried with integrated Intel GPU and NVidea gfx. So far the issue does not show on my system. How long should it take after startup until that happens?

**Comment by afedchin** [comment]
Strange. It happens after a several selection different home menu items. It also happens randomly on video list, select dialog

-----Original Message-----

**Comment by MartijnKaijser** [comment]
seeing it too
https://dl.dropboxusercontent.com/u/7435592/screenshot018.png

**Comment by FernetMenta** [comment]
ok, I managed it to see something similar as @MartijnKaijser but I doubt it is the same issue as shown in screenshot by Montellese. I see the text of list items disappear completely. And it comes back when navigating some items up/down

**Comment by ace20022** [comment]
The montellese thing coulb be reproduced by resizing the window (with kodi master)

**Comment by FernetMenta** [comment]
@afedchin resizing with your branch is broken. there is no rendering while the window is resized.

**Comment by afedchin** [comment]
@FernetMenta do you mean after resize there is no rendering or during resize the picture is freezed

**Comment by FernetMenta** [comment]
it's not an issue of your branch. I set back to master and same issue. I mean during resize rendering should continue with updated window size.

**Comment by Karlson2k** [comment]
@FernetMenta Do you mean **after** resize rendering must continue?

**Comment by FernetMenta** [comment]
rendering should not stop. if you have a Linux system, try how it works there. if not, use WMP. There is continuous rendering WHILE resizing. 
currently on Windows if you resize to the smallest size, keep left mouse button pressed, and enlarge window again: it stays black until you release the button.

**Comment by afedchin** [comment]
confirm. rendering stops. just press and _hold_ mouse button at the border of window. It is a windows "feature". in short: when DefWindowProc handles WM_SYSCOMMAND with either SC_MOVE or SC_SIZE in the wParam, it enters a loop until the user stops it by releasing the mouse button, or pressing either enter or escape. There is a little hack to resolve this. I'll try it tomorrow.

**Comment by afedchin** [comment]
I've found reason of disappearing font and fixed it in dx11 implementation. But there are a couple other issues: trying to create an empty vertex buffer (not issue but still). Other issue is the static vertext buffers not released. debug layer reports about 300 unreleased buffers on exit and their count is greater if Kodi works longer.

File: xbmc/guilib/GUIFontCache.h [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+#include <stdint.h>
+
+#include <algorithm>
+#include <vector>
+
+#include "boost/multi_index_container.hpp"
+#include "boost/multi_index/sequenced_index.hpp"
+#include "boost/multi_index/hashed_index.hpp"
+#include "boost/multi_index/member.hpp"
+#include "boost/shared_ptr.hpp"
```
Comment (Memphiz) [inline xbmc/guilib/GUIFontCache.h]: @bavison well i know i might come over annoying - but latest xcode which was released a couple of days ago chokes on those includes again. It seems the boost version we are using here on osx (1.44) doesn't really play well with c++11. Just fyi here is the printout ( you can't do anything about it - the problem is burried in one of the boost headers which gets implicitly included):

http://pastebin.com/ddQgCmdW

Can you think of any approach which could refactor this stuff without the usage of this boost stuff? ATM its not needed to support xcode 6.3 - but users will come and ask for support because it doesn't compile with latest osx build tools at some point ...

Comment (Memphiz) [inline xbmc/guilib/GUIFontCache.h]: mhh i might have found a workaround for it ... so hold back for now ;)

**Comment by FernetMenta** [comment]
@bavison @popcornmix this change turned out to cause issues with text borders of subtitles: http://trac.kodi.tv/ticket/15971
any ideas?

**Comment by arnova** [comment]
FYI: Seems like this is the same issue I mentioned in the "Isengard issue list".

**Comment by MilhouseVH** [comment]
I suspect (this)[http://forum.kodi.tv/showthread.php?tid=227071] is the same issue, on RPi. Will try and confirm later today.

**Comment by MilhouseVH** [comment]
With an RPi2 using my latest #0526 test build I can see the font corruption when using the test.srt file from [trac #15971](http://trac.kodi.tv/ticket/15971#comment:5), but only with the 4th of the 5 subtitles:

screenshot

Reverting this PR is almost impossible however, as it dates back to Feb 2014 on RPi.

I did go back to some really old builds, and [#0202(2014)](http://forum.kodi.tv/showthread.php?tid=184866&pid=1617155#pid1617155) is the first build that shows font glitch issues when rendering test.srt - this build includes:
```
newclock3: Ben Avison patches that should help with text rendering speed
```
Unfortunately the github comparison for this build is not longer functioning, so I'm not able to view precisely which commits were included from the newclock3 branch.

The build immediately prior to 0202, (#0201)[http://forum.kodi.tv/showthread.php?tid=184866&pid=1616399#pid1616399] has no problem displaying test.srt.

These old builds can be downloaded from an [archive folder](http://netlir.dk/rbej/builds/MilhouseVH/archive/Gotham)

**Comment by Memphiz** [comment]
I saw the issue on a helix beta build on rbpi2 (not sure if this build contains newclock3 but its a build from oe mainline without any additions...)

**Comment by FernetMenta** [comment]
definitely this change. I disabled the caching as shown here https://github.com/bavison/xbmc/commit/5020a87ad21bc60fa36500312c3d0363444e0d47
then the issue disappeared

**Comment by afedchin** [comment]
I can confirm on dx11 build which has this. Disabling caching with https://github.com/bavison/xbmc/commit/5020a87ad21bc60fa36500312c3d0363444e0d47 dissapears the issue.

**Comment by bavison** [comment]
Looks like this patch series simply made a pre-existing bug more visible. I've written a fix over at PR #7201.
